feat#23/user revoke

kkyu0718 opened the pull request:
## ✅ PR 체크
- [ ] 하나의 PR 에는 100줄 정도의 커밋을 한다는 규칙을 지키고 있나요?
- [x] 무엇을 변경했는지 충분히 설명하고 있나요?
- [x] 새로운 기술을 사용했다면, 그 기술을 설명하고 있나요?
- [x] 이해하기 어려운 비즈니스 로직에 관해서 부연 설명을 하고 있나요?
- [x] 팀원 한 명을 Assign 했나요?
- [x] 작업 범위에 대해서 테스트를 작성하셨나요?

## 🚨 관련 이슈
closes #23

## 🌈 작업 상황
- 현준님이랑 conflict 가 나는 상황이라 브랜치를 새로 팠습니다.
- 유저 도메인 회원가입 회원탈퇴 로직 작성
- 테스트 코드 작성

## 📌 기타
- ApiResponse 가 200으로만 상태코드를 던지는 상황 해결 필요

kkyu0718 committed:
feat: MemberCreateDto 를 Member 엔티티로 변환하는 메소드 추가

Jpa Entity 로 변환하기 위한 toEntity 메소드 작성

kkyu0718 committed:
fix: querydsl build.gradle 설정 수정

querydsl 설정이 springboot 3.x 부터 달라져 생긴 QClass 생성 오류 해결

참고 https://lordofkangs.tistory.com/461

Commit:
refactor: memberName을 name으로 변수명 변경

- name으로 하고 테이블 명으로 구분해줬습니다.

Commit:
build: jacoco 설정 git conflict 수정

- 제외 클래스, 패키지 추가

Commit:
ci: Github Action 실행 환경을 Docker로 변경

- DB가 없어 테스트를 수행할 수 없어서 도커로 환경 설정

Commit:
chore: Github Action에 깔린 도커로 변경

Commit:
build: 빌드 시 h2 데이터베이스 환경으로 테스트가 돌아 갈 수 있도록 수정

- 도커는 로컬로만 쓰고 h2로 빌드하도록 수정

kkyu0718 committed:
feat: UserErrorCode 클래스 추가

유저 도메인에 한하여 커스텀 Exception 를 사용하기 위해 이용할 에러 코드 정의

kkyu0718 committed:
fix: 테스트 코드 예외처리 핸들러 추가

WebMvcTest 에서 핸들러를 타지않아 그대로 에러가 올라오는 오류가 있어 수정

mockMvc 에 exception handler 직접 추가

kkyu0718 committed:
feat: 유저 회원탈퇴 로직 구현

soft delete 되도록 Member 도메인에 isDeleted=True 로 변경하는 메소드 추가

kkyu0718 committed:
feat: 회원 탈퇴 서비스 로직 구현

영속성 컨텍스트를 사용하기 위해 @Transactional 사용

isDeleted=True 로 변경하여 flush 하도록 함

kkyu0718 committed:
test: 회원 탈퇴 로직 테스트 추가

- 회원 탈퇴를 한 유저는 조회되어야하고 isDeleted=True 이어야함
- 회원 탈퇴를 요청한 유저는 db 에서 유효한 유저여야함

kkyu0718 committed:
test: 회원 탈퇴 로직 테스트 추가

- 회원 탈퇴를 한 유저는 조회되어야하고 isDeleted=True 이어야함
- 회원 탈퇴를 요청한 유저는 db 에서 유효한 유저여야함

kkyu0718 committed:
feat: 회원탈퇴 컨트롤러 코드 작성

세션에서 memberId 를 가져와서 db 에 접근

member 의 isDeleted=True 로 수정

kkyu0718 committed:
refactor: ResponseEntity 를 ApiResponse 로 변경

ApiResponse 클래스가 추가됨에 따라 ApiResponse 로 응답 변경

kkyu0718 committed:
fix: 컨트롤러 테스트 성공 상태코드 모두 200으로 수정

ApiResponse 공통 클래스를 사용하다보니 모두 200으로 반환되어 테스트 코드 성공 상태코드를 수정

추후에 수정 필요

kkyu0718 committed:
fix: ApiExceptionHandler 우선순위 제일 높게 수정

우선순위 때문에 GlobalExceptionHandler 를 먼저 타게 되어 500 에러로 모두 던져지는 상황 발생

우선순위를 제일 높게 주어 먼저 수행되도록 수정

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/domain/member/controller/MemberController.java:
```diff
@@ -19,7 +21,19 @@ public class MemberController {
+    기능 확장성을 고려하면 회원 탈퇴할 유저 id 를 받아야함
+     */
+    @DeleteMapping
+    public ApiResponse<MemberIdDto> revokeMember(HttpSession session) {
```
revoke보다는 다른 이름이 어떨까요? 취소 보다는 삭제라는 의미를 담은 delete가 문맥 상 더 이해하기 쉬울 거 같아요 !

kkyu0718 replied:
delete 는 db 에서 완전히 지운다는 것으로 오해될거 같아서 쓰지 않았습니다!
철회라는 의미에서 revoke 를 쓴거긴한데 탈퇴이면 quit 은 어떠신가요?

HyeonJun0530 replied:
delete를 soft delete로 하기로 합의하고 이미 멤버 도메인 메서드에 상태를 업데이트하는 로직이름에 delete라는 이름이 포함되어 있어서 문제 없을 거 같아요 😊

kkyu0718 replied:
withdraw 로 바꾸기로 논의됨

HyeonJun0530 commented on src/main/java/com/wootecam/festivals/domain/member/exception/UserErrorCode.java:
```diff
@@ -0,0 +1,33 @@
+        this.message = message;
+    }
+
+    @Override
+    public String getName() {
+        return name();
+    }
+
+    @Override
+    public String getDescription() {
+        return httpStatus + ": " + code + " - " + message;
+    }
+}
```
이 부분은 중복 부분인데 추후 디폴트 메서드 같은 것으로 바꿔야 할지에 대해 논의 해봐야 할 필요가 있는 것 같습니다 !

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/member/service/MemberService.java:
```diff
@@ -15,11 +19,19 @@ public Long createMember(MemberCreateDto memberCreateDto) {
+    public void revokeMember(Long memberId) {
+        Member member = memberRepository.findById(memberId)
+                .orElseThrow(() -> new ApiException(UserErrorCode.USER_NOT_FOUND));
+
```
p1: 해당 회원이 organization의 유일한 관리자인 경우, 관리자 권한을 다른 사람에게 양도해야 하니 실패하는 로직이 필요할 것 같습니다!

kkyu0718 replied:
OrganizationMember 테이블을 보는 로직이 필요하겠군요! 그럼 OrganizationMemberRepository와 MemberRepository 둘다 필요한 상황인데 이걸 어디 서비스 클래스에 넣을지 고민되네요! (오전에 이야기했던 것의 연장선!)

HyeonJun0530 reviewed (commented):
API에 대한 asciidoc도 작성해주시면 좋을 것 같아요 !
궁금한 점, 논의해야 할 점은 리뷰에 남겼습니다 !

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/member/service/MemberService.java:
```diff
@@ -15,11 +19,19 @@ public Long createMember(MemberCreateDto memberCreateDto) {
+        Member member = memberRepository.findById(memberId)
+                .orElseThrow(() -> new ApiException(UserErrorCode.USER_NOT_FOUND));
+
+        member.updateStatusDeleted();
```
p3: email, name 등 개인 정보가 더 이상 필요하지 않으니 스크럼블하는 로직이 필요...할 것 같으나 다시 복귀할 수도 있으니 배치 처리하는 것이 더 올바를 수도 있겠네요. pr에 올려주실 수 있으실까요?

kkyu0718 replied:
스크럼블한다는 것이 어떠한 의미죠?!

minji-park-eng replied:
닉네임 user1과 같은 기존 데이터를 deleted-user와 같이 뭉갠다는 소리였습니다!

minji-park-eng commented on src/main/java/com/wootecam/festivals/domain/member/exception/UserErrorCode.java:
```diff
@@ -0,0 +1,33 @@
+
+@Getter
+public enum UserErrorCode implements ErrorCode, EnumType {
+    USER_NOT_FOUND(HttpStatus.NOT_FOUND, "U-0001", "사용자를 찾을 수 없습니다."),
```
p1: 클래스 정의 밑 개행 컨벤션! 개행 부탁드립니다! :)

minji-park-eng commented on src/test/java/com/wootecam/festivals/domain/member/service/MemberServiceTest.java:
```diff
@@ -62,7 +68,41 @@ void createMemberWithDuplicatedEmail() {
 
         // when, then
-        assertThrows(IllegalArgumentException.class,
+        ApiException ex = assertThrows(ApiException.class,
                 () -> memberService.createMember(memberCreateDto));
+        assertEquals(ex.getErrorCode(), UserErrorCode.DUPLICATED_EMAIL);
```
p3: assertJ의 hasMessage()로 메시지까지 한번에 검증할 수 있습니다!

minji-park-eng reviewed (commented):
코드 잘 봤습니다! 회원 탈퇴 로직... 제일 어려운 부분이라 나중에 수정할 부분이 있을 것 같아요! 고생하셨습니다!

kkyu0718 committed:
refactor: Dto Response 와 Request 가 구별되도록 클래스명 변경

- MemberIdDto -> MemberResponseDto
- MemberCreateDto -> MemberCreateRequestDto

kkyu0718 committed:
refactor: 회원탈퇴 메소드명 변경

컨트롤러와 서비스의 회원탈퇴 메소드명 변경 revokeMember -> withdrawMember

탈퇴의 의미가 더 명확해지도록 메소드명 변경

kkyu0718 committed:
refactor: 회원탈퇴 메소드명 변경

테스트 코드의 회원탈퇴 메소드명 변경 revokeMember -> withdrawMember

탈퇴의 의미가 더 명확해지도록 메소드명 변경

kkyu0718 committed:
refactor: 회원탈퇴 메소드명 변경

테스트 코드의 회원탈퇴 메소드명 변경 revokeMember -> withdrawMember

탈퇴의 의미가 더 명확해지도록 메소드명 변경

kkyu0718 committed:
docs: member adoc 문서 작성

kkyu0718 committed:
style: UserErrorCode 개행문자 추가

코드 컨벤션 따르기 위해 클래스 정의 후 한줄 개행

kkyu0718 committed:
refactor: assertj 의존성 추가 후 exception 테스트 재작성

유저 회원가입 시 중복된 이메일 에러 로직을 더 편리하게 테스트하기 위한 assertj 의존성 추가 후 테스트코드 재작성

kkyu0718 committed:
refactor: 유저 회원가입 상태코드 201로 변경

유저 회원가입 시 200 상태코드로 나가던 것을 201 로 변경

kkyu0718 committed:
Merge branch 'develop' into feat#23/user-revoke

minji-park-eng reviewed (approved):
고생하셨습니다!!
